src/modules/silentpayments/main_impl.h

```diff
@@ -7,10 +7,233 @@
+    int overflow = 0;
+    int ret = 1;
+
+    /* Sanity check inputs. */
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));
+    ARG_CHECK(generated_outputs != NULL);
+    ARG_CHECK(recipients != NULL);
+    ARG_CHECK(n_recipients > 0);
+    ARG_CHECK((plain_seckeys != NULL) || (taproot_seckeys != NULL));
+    if (taproot_seckeys != NULL) {
+        ARG_CHECK(n_taproot_seckeys > 0);
+    } else {
+        ARG_CHECK(n_taproot_seckeys == 0);
+    }
+    if (plain_seckeys != NULL) {
+        ARG_CHECK(n_plain_seckeys > 0);
+    } else {
+        ARG_CHECK(n_plain_seckeys == 0);
+    }
+    ARG_CHECK(outpoint_smallest36 != NULL);
+    /* ensure the index field is set correctly */
+    for (i = 0; i < n_recipients; i++) {
+        ARG_CHECK(recipients[i]->index == i);
+    }
+
+    /* Compute input private keys sum: a_sum = a_1 + a_2 + ... + a_n */
+    a_sum_scalar = secp256k1_scalar_zero;
+    for (i = 0; i < n_plain_seckeys; i++) {
+        /* TODO: in other places where _set_b32_seckey is called, its normally followed by a _cmov call
+         * Do we need that here and if so, is it better to call it after the loop is finished?
+         */
+        ret &= secp256k1_scalar_set_b32_seckey(&addend, plain_seckeys[i]);
+        secp256k1_scalar_add(&a_sum_scalar, &a_sum_scalar, &addend);
+    }
+    /* private keys used for taproot outputs have to be negated if they resulted in an odd point */
+    for (i = 0; i < n_taproot_seckeys; i++) {
+        secp256k1_ge addend_point;
+        /* TODO: why don't we need _cmov here after calling keypair_load? Because the ret is declassified? */
+        ret &= secp256k1_keypair_load(ctx, &addend, &addend_point, taproot_seckeys[i]);
+        if (secp256k1_fe_is_odd(&addend_point.y)) {
+            secp256k1_scalar_negate(&addend, &addend);
+        }
+        secp256k1_scalar_add(&a_sum_scalar, &a_sum_scalar, &addend);
+    }
+    /* If there are any failures in loading/summing up the secret keys, fail early */
+    if (!ret || secp256k1_scalar_is_zero(&a_sum_scalar)) {
+        return 0;
+    }
+    /* Compute input_hash = hash(outpoint_L || (a_sum * G)) */
+    secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &A_sum_gej, &a_sum_scalar);
+    secp256k1_ge_set_gej(&A_sum_ge, &A_sum_gej);
+
+    /* Calculate the input hash and tweak a_sum, i.e., a_sum_tweaked = a_sum * input_hash */
+    secp256k1_silentpayments_calculate_input_hash(input_hash, outpoint_smallest36, &A_sum_ge);
+    secp256k1_scalar_set_b32(&input_hash_scalar, input_hash, &overflow);
+    ret &= !overflow;
+    /* TODO: any concerns here with multiplying a "secret" by a scalar here, i.e., input_hash * a_sum
+     * Mostly concerned about cases where a_sum is a single key and leaking information about a_sum
+     * since input_hash can be calculated by anyone
+     */
+    secp256k1_scalar_mul(&a_sum_scalar, &a_sum_scalar, &input_hash_scalar);
```

Comment:
But if `a_sum` is a single key, is it possible to calculate the secret key, given any derived public information (such as the generated output) ?

Reply:
> I strongly assume that this is fine

Same. Left the comment more to draw eyes from the "Friendly Neighborhood Cryptographers." I'm not aware of any timing attacks that work with multiplying scalars or information that could be leaked, but wanted to double check.

> is it possible to calculate the secret key, given any derived public information

No. This is a question about whether or not an attacker can learn anything from the multiplication step.

Reply:
Removed the comment. I left a few of the other TODO comments in regarding `_cmov` and loading unsigned char secret keys into scalars because that's something I'd appreciate some eyes on / feedback.